### PR TITLE
AWS agents: reuse IMDS token until it expires (issue #1990)

### PR DESCRIPTION
**Pull request by harshkiprofile**

### CODE REVIEW

File: heartbeat/ocf-shellfuncs.in [hunk 1]
```diff
@@ -1111,3 +1111,34 @@ ocf_is_true "$OCF_TRACE_RA" && ocf_start_trace
...
+# Function to fetch a new token
+fetch_new_token() {
+    TOKEN=$(curl -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: $TOKEN_LIFETIME")
+    echo "$TOKEN $(date +%s)" > "$TOKEN_FILE"
+    echo "$TOKEN"
+}
+
+# Function to retrieve or renew the token
+get_token() {
+    if [[ -f "$TOKEN_FILE" ]]; then
```
Comment (oalbrigt) [inline heartbeat/ocf-shellfuncs.in:1129]: Use single square brackets here to avoid depending on bash for no reason.

**Review by oalbrigt — CHANGES_REQUESTED** [review]
Added comments that the new functions needs to be moved.

get_token() should be used in the other AWS agents as well.

**Comment by harshkiprofile** [comment]
Thanks @oalbrigt - Changes done. Please re-review

File: heartbeat/Makefile.am [hunk 2]
```diff
@@ -218,6 +218,7 @@ ocfcommon_DATA		= ocf-shellfuncs 	\
 			  ocf-rarun		\
 			  ocf-distro		\
 			  apache-conf.sh 	\
+			  aws.sh            \
```
Comment (oalbrigt) [inline heartbeat/Makefile.am:221]: You should use tabs after aws.sh as well, so `\` lines up with the other lines.

You can run `git diff` for a preview of how it looks outside your editor.

File: heartbeat/aws.sh [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+#
+#
+#   AWS Helper Scripts
+#
+#
+
+: ${OCF_FUNCTIONS_DIR=${OCF_ROOT}/lib/heartbeat}
+. ${OCF_FUNCTIONS_DIR}/ocf-shellfuncs
+
+# Defaults
```
Comment (oalbrigt) [inline heartbeat/aws.sh:11]: Good thinking. You just need to remove _default lines for curl_retries/curl_sleep from the agent as well.

File: heartbeat/aws.sh [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+# Function to retrieve or renew the token
+get_token() {
+    if [ -f "$TOKEN_FILE" ]; then
+        read -r STORED_TOKEN STORED_TIMESTAMP < "$TOKEN_FILE"
+        CURRENT_TIME=$(date +%s)
+        ELAPSED_TIME=$((CURRENT_TIME - STORED_TIMESTAMP))
+
+        if (( ELAPSED_TIME < (TOKEN_LIFETIME - TOKEN_EXPIRY_THRESHOLD) )); then
```
Comment (oalbrigt) [inline heartbeat/aws.sh:38]: This seems to be required bash, so you should use e.g. `if [ "$ELAPSED_TIME" -lt "$((TOKEN_LIFETIME - TOKEN_EXPIRY_THRESHOLD))" ]; then`.

**Review by oalbrigt — CHANGES_REQUESTED** [review]
Looking good.

I've added some suggestions for improvements.

File: heartbeat/aws.sh (hunk 3 above)
Comment (harshkiprofile) [reply heartbeat/aws.sh:11]: Thanks @oalbrigt - I believe keeping this variable in the agent is still valuable, as it’s used in other metadata calls. However, if you feel it's better to remove it, please let me know, and I’m happy to make the change.

image

Comment (oalbrigt) [reply heartbeat/aws.sh:11]: I agree, but it should only be in one place. So either in the agents or in `aws.sh`.

Comment (harshkiprofile) [reply heartbeat/aws.sh:11]: That makes sense. Thanks for the feedback!

File: heartbeat/aws.sh [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+#
+
+: ${OCF_FUNCTIONS_DIR=${OCF_ROOT}/lib/heartbeat}
+. ${OCF_FUNCTIONS_DIR}/ocf-shellfuncs
+
+# Defaults
+OCF_RESKEY_curl_retries_default="3"
+OCF_RESKEY_curl_sleep_default="1"
+
+: ${OCF_RESKEY_curl_retries=${OCF_RESKEY_curl_retries_default}}
```
Comment (oalbrigt) [inline heartbeat/aws.sh:15]: You also need to remove these lines from the agents.

**Comment by oalbrigt** [comment]
`aws.sh` needs to added to be `symlinkstargets` in `doc/man/Makefile.am`.

File: doc/man/Makefile.am [hunk 6]
```diff
@@ -43,7 +43,8 @@ radir			= $(abs_top_builddir)/heartbeat
 symlinkstargets		= \
 			  ocf-distro ocf.py ocf-rarun ocf-returncodes \
 			  findif.sh apache-conf.sh http-mon.sh mysql-common.sh \
-			  nfsserver-redhat.sh openstack-common.sh ora-common.sh
+			  nfsserver-redhat.sh openstack-common.sh ora-common.sh \
+                          aws.sh
```
[the remark this answers is not in the document]
Comment (harshkiprofile) [reply doc/man/Makefile.am]: Should be good now

**Review by oalbrigt — APPROVED** [review]

**Comment by oalbrigt** [comment]
Looks good to me. I'll merge it tomorrow.

**Comment by oalbrigt** [comment]
Thanks.

**Comment by ebilbaoANZ** [comment]
Hi team, thanks for your work here. We still have a risk with this logic due to values of: OCF_RESKEY_curl_retries_default="3" and OCF_RESKEY_curl_sleep_default="1" in the new _aws.sh_.

The AWS IMDS token service can be unavailable longer than 3 seconds. The sleep timer / retry loop has to exceed 10 seconds in it's request for a token. 

I suggest we change the default settings here to:
OCF_RESKEY_curl_retries_default="4" and OCF_RESKEY_curl_sleep_default="5"

**Pull request merged**